rllib/utils/postprocessing/zero_padding.py

```diff
@@ -20,123 +20,150 @@ def create_mask_and_seq_lens(
+            BatchedNdArray,
+            split_and_zero_pad,
+        )
+        from ray.rllib.utils.test_utils import check
 
-            # Pad the chunk if it's shorter than T
-            if sub_chunk.shape[0] < T:
-                padding_shape = [(0, T - sub_chunk.shape[0])] + [
-                    (0, 0) for _ in range(sub_chunk.ndim - 1)
-                ]
-                sub_chunk = np.pad(sub_chunk, pad_width=padding_shape, mode="constant")
+        # Simple case: `item_list` contains individual floats.
+        check(
+            split_and_zero_pad([0, 1, 2, 3, 4, 5, 6, 7], 5),
+            [[0, 1, 2, 3, 4], [5, 6, 7, 0, 0]],
+        )
 
-            all_chunks.append(sub_chunk)
+        # `item_list` contains BatchedNdArray (ndarrays that explicitly declare they
+        # have a batch axis=0).
+        check(
+            split_and_zero_pad([
+                BatchedNdArray([0, 1]),
+                BatchedNdArray([2, 3, 4, 5]),
+                BatchedNdArray([6, 7, 8]),
+            ], 5),
+            [[0, 1, 2, 3, 4], [5, 6, 7, 8, 0]],
+        )
 
-    # Combine all chunks into a single array.
-    result = np.concatenate(all_chunks, axis=0)
+    Args:
+        item_list: A list of individual items or BatchedNdArrays to be split into
+            `max_seq_len` long pieces (the last of which may be zero-padded).
+        max_seq_len: The maximum length of each item in the returned list.
 
-    # Reshape the array to include the time dimension T.
-    # The new shape should be (-1, T) + original dimensions (excluding the
-    # batch dimension).
-    result = result.reshape((-1, T) + result.shape[1:])
+    Returns:
+        A list of np.ndarrays (all of length `max_seq_len`), which contains the same
+        data as `item_list`, but split into sub-chunks of size `max_seq_len`.
+        The last item in the returned list may be zero-padded, if necessary.
+    """
+    zero_element = tree.map_structure(
+        lambda s: np.zeros_like([s[0]] if isinstance(s, BatchedNdArray) else s),
+        item_list[0],
+    )
+
+    # The replacement list (to be returned) for `items_list`.
+    # Items list contains n individual items.
+    # -> ret will contain m batched rows, where m == n // T and the last row
+    # may be zero padded (until T).
+    ret = []
+
+    # List of the T-axis item, collected to form the next row.
+    current_time_row = []
+    current_t = 0
+
+    item_list = deque(item_list)
+    while len(item_list) > 0:
+        item = item_list.popleft()
+        # `item` is already a batched np.array: Split if necessary.
+        if isinstance(item, BatchedNdArray):
+            t = max_seq_len - current_t
+            current_time_row.append(item[:t])
+            if len(item) <= t:
+                current_t += len(item)
+            else:
+                current_t += t
+                item_list.appendleft(item[t:])
+        # `item` is a single item (no batch axis): Append and continue with next item.
```

Comment:
Could it happen that `item` has no batch axis, but a time axis (for example with `max_seq_len` time axis length)?

Reply:
Yeah, this is confusing. So if `item` has a "batch" axis, then that axis (axis=0) is assumed to be the time axis (of a coherent trajectory). So item is either a single item (without batch/time axes) OR a sequence of items (with a time-axis, which we'll call "batch" axis here).

You are right, maybe we should rename this here in this function.
On the other hand, this would take the generality away from our `batch` utility function, which is used for both actual batch and time contexts.